clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java

```diff
@@ -280,26 +417,70 @@ public void appendUnchecked(long offset, Record record) {
     }
 
     /**
-     * Add a record with a given offset. The record must have a magic which matches the magic use to
-     * construct this builder and the offset must be greater than the last appended entry.
+     * Append a record at the next sequential offset.
+     * @param record the record to add
+     */
+    public void append(Record record) {
+        appendWithOffset(record.offset(), record.isControlRecord(), record.timestamp(), record.key(), record.value(),
```

Comment:
This method is only used by `ByteBufferMessageSet` (i.e. Scala clients) so it seems OK to favour consistency over optimisation in this case.